File: src/main/java/destiny/core/calendar/JulianDateTime.java
/**
 * Created by smallufo on 2017-01-20.
 */
package destiny.core.calendar;

import kotlin.Pair;
import org.threeten.extra.chrono.JulianDate;

import java.io.Serializable;
import java.time.*;
import java.time.chrono.ChronoLocalDateTime;
import java.time.temporal.*;
import java.util.Objects;

import static java.time.temporal.ChronoField.*;

/**
 * reference : {@link java.time.LocalDateTime}
 */
public class JulianDateTime implements Serializable , ChronoLocalDateTime<JulianDate> {

  /**
   * Hours per day.
   */
  private static final int HOURS_PER_DAY = 24;

  /**
   * Minutes per hour.
   */
  private static final int MINUTES_PER_HOUR = 60;

  /**
   * Minutes per day.
   */
  private static final int MINUTES_PER_DAY = MINUTES_PER_HOUR * HOURS_PER_DAY;

  /**
   * Seconds per minute.
   */
  private static final int SECONDS_PER_MINUTE = 60;

  /**
   * Seconds per hour.
   */
  private static final int SECONDS_PER_HOUR = SECONDS_PER_MINUTE * MINUTES_PER_HOUR;

  /**
   * Seconds per day.
   */
  private static final int SECONDS_PER_DAY = SECONDS_PER_HOUR * HOURS_PER_DAY;

  /**
   * Milliseconds per day.
   */
  private static final long MILLIS_PER_DAY = SECONDS_PER_DAY * 1000L;

  /**
   * Microseconds per day.
   */
  private static final long MICROS_PER_DAY = SECONDS_PER_DAY * 1000_000L;

  /**
   * Nanos per second.
   */
  private static final long NANOS_PER_SECOND = 1000_000_000L;

  /**
   * Nanos per minute.
   */
  private static final long NANOS_PER_MINUTE = NANOS_PER_SECOND * SECONDS_PER_MINUTE;

  /**
   * Nanos per hour.
   */
  private static final long NANOS_PER_HOUR = NANOS_PER_MINUTE * MINUTES_PER_HOUR;

  /**
   * Nanos per day.
   */
  private static final long NANOS_PER_DAY = NANOS_PER_HOUR * HOURS_PER_DAY;

  private final JulianDate date;
  private final LocalTime time;


  private JulianDateTime(JulianDate date, LocalTime time) {
    this.date = date;
    this.time = time;
  }

  public static JulianDateTime from(TemporalAccessor temporal) {
    if (temporal instanceof JulianDateTime) {
      return (JulianDateTime) temporal;
    }
    else if (temporal instanceof ZonedJulianDateTime) {
      return ((ZonedJulianDateTime) temporal).toLocalDateTime();
    }
//    else if (temporal instanceof OffsetDateTime) {
//      return ((OffsetDateTime) temporal).toLocalDateTime();
//    }
    try {
      JulianDate date = JulianDate.from(temporal);
      LocalTime time = LocalTime.from(temporal);
      return new JulianDateTime(date, time);
    } catch (DateTimeException ex) {
      throw new DateTimeException("Unable to obtain JulianDateTime from TemporalAccessor: " + temporal + " of type " + temporal.getClass().getName(), ex);
    }
  }

  public static JulianDateTime of(JulianDate date, LocalTime time) {
    Objects.requireNonNull(date, "date");
    Objects.requireNonNull(time, "time");
    return new JulianDateTime(date, time);
  }

  /**
   * @param prolepticYear maybe <= 0
   */
  public static JulianDateTime of(int prolepticYear, int month, int dayOfMonth, int hour, int minute , int second , int nano) {
    JulianDate date = JulianDate.of(prolepticYear , month , dayOfMonth);
    LocalTime time = LocalTime.of(hour , minute , second , nano);
    return new JulianDateTime(date , time);
  }

  /**
   * @param prolepticYear maybe <= 0
   */
  public static JulianDateTime of(int prolepticYear, int month, int dayOfMonth, int hour, int minute , int second) {
    JulianDate date = JulianDate.of(prolepticYear , month , dayOfMonth);

    Pair<Integer , Integer> pair = TimeTools.Companion.splitSecond(second);
    LocalTime time = LocalTime.of(hour , minute , pair.getFirst(), pair.getSecond());
    return new JulianDateTime(date , time);
  }

  /**
   * @param prolepticYear maybe <= 0
   */
  public static JulianDateTime of(int prolepticYear, int month, int dayOfMonth, int hour, int minute) {
    return of(prolepticYear , month , dayOfMonth , hour , minute , 0);
  }

  /**
   * @param prolepticYear maybe <= 0
   */
  public static JulianDateTime of(int prolepticYear, Month month, int dayOfMonth, int hour, int minute) {
    return of(prolepticYear , month.getValue() , dayOfMonth , hour , minute , 0);
  }

  public static JulianDateTime ofEpochSecond(long epochSecond, int nanoOfSecond, ZoneOffset offset) {
    Objects.requireNonNull(offset, "offset");
    NANO_OF_SECOND.checkValidValue(nanoOfSecond);
    long localSecond = epochSecond + offset.getTotalSeconds();  // overflow caught later
    long localEpochDay = Math.floorDiv(localSecond, (long)SECONDS_PER_DAY);
    int secsOfDay = (int) Math.floorMod(localSecond, (long)SECONDS_PER_DAY);
    LocalDate gDate = LocalDate.ofEpochDay(localEpochDay);
    JulianDate date = JulianDate.from(gDate);
    LocalTime time = LocalTime.ofNanoOfDay(secsOfDay * NANOS_PER_SECOND + nanoOfSecond);
    return new JulianDateTime(date, time);
  }

  /** 一定為正值 */
  public int getYearOfEra() {
    return date.get(YEAR_OF_ERA);
  }

  /** 與 {@link LocalDate#year} 一樣，都為 proleptic year . 西元元年=1 , 西元前一年=0 */
  public int getYear() {
    return date.get(YEAR);
  }

  public int getMonth() {
    return date.get(MONTH_OF_YEAR);
  }

  public int getDayOfMonth() {
    return date.get(DAY_OF_MONTH);
  }

  public int getHour() {
    return time.getHour();
  }

  public int getMinute() {
    return time.getMinute();
  }

  public int getSecond() {
    return time.getSecond();
  }

  public int getNano() {
        return time.getNano();
    }

  public JulianDate toLocalDate() {
    return date;
  }

  public LocalTime toLocalTime() {
    return time;
  }

  /**
   * see {@link java.time.LocalDateTime#isSupported(TemporalField)}
   */
  @Override
  public boolean isSupported(TemporalField field) {
    if (field instanceof ChronoField) {
      ChronoField f = (ChronoField) field;
      return f.isDateBased() || f.isTimeBased();
    }
    return field != null && field.isSupportedBy(this);
  }

  /**
   * see {@link java.time.LocalDateTime#getLong(TemporalField)}
   */
  @Override
  public long getLong(TemporalField field) {
    if (field instanceof ChronoField) {
      ChronoField f = (ChronoField) field;
      return (f.isTimeBased() ? time.getLong(field) : date.getLong(field));
    }
    return field.getFrom(this);
  }

  /**
   * see {@link LocalDateTime#with(TemporalField, long)}
   */
  @Override
  public ChronoLocalDateTime<JulianDate> with(TemporalField field, long newValue) {
    if (field instanceof ChronoField) {
      ChronoField f = (ChronoField) field;
      if (f.isTimeBased()) {
        return with(date, time.with(field, newValue));
      }
      else {
        return with(date.with(field, newValue), time);
      }
    }
    return field.adjustInto(this, newValue);
  }

  /**
   * see {@link LocalDateTime#with(LocalDate, LocalTime)}
   */
  private JulianDateTime with(JulianDate newDate, LocalTime newTime) {
    if (date == newDate && time == newTime) {
      return this;
    }
    return new JulianDateTime(newDate, newTime);
  }

  /**
   * see {@link LocalDateTime#plus(long, TemporalUnit)}
   */
  @Override
  public ChronoLocalDateTime<JulianDate> plus(long amountToAdd, TemporalUnit unit) {
    if (unit instanceof ChronoUnit) {
      ChronoUnit f = (ChronoUnit) unit;
      switch (f) {
        case NANOS: return plusNanos(amountToAdd);
        case MICROS:return plusDays(amountToAdd / MICROS_PER_DAY).plusNanos((amountToAdd % MICROS_PER_DAY) * 1000);
        case MILLIS:return plusDays(amountToAdd / MILLIS_PER_DAY).plusNanos((amountToAdd % MILLIS_PER_DAY) * 1000_000);
        case SECONDS:return plusSeconds(amountToAdd);
        case MINUTES:return plusMinutes(amountToAdd);
        case HOURS:return plusHours(amountToAdd);
        case HALF_DAYS:return plusDays(amountToAdd / 256).plusHours((amountToAdd % 256) * 12);  // no overflow (256 is multiple of 2)
      }
      return with(date.plus(amountToAdd, unit), time);
    }
    return unit.addTo(this, amountToAdd);
  }

  @Override
  public ChronoLocalDateTime<JulianDate> minus(long amountToSubtract, TemporalUnit unit) {
    return plus(-amountToSubtract , unit);
  }

  private JulianDateTime plusNanos(long nanos) {
    return plusWithOverflow(date, 0, 0, 0, nanos, 1);
  }

  private JulianDateTime plusSeconds(long seconds) {
    return plusWithOverflow(date, 0, 0, seconds, 0, 1);
  }

  private JulianDateTime plusMinutes(long minutes) {
    return plusWithOverflow(date, 0, minutes, 0, 0, 1);
  }

  private JulianDateTime plusHours(long hours) {
    return plusWithOverflow(date, hours, 0, 0, 0, 1);
  }

  private JulianDateTime plusDays(long days) {
    JulianDate newDate = date.plus(days , ChronoUnit.DAYS);
    return with(newDate, time);
  }

  /**
   * see {@link LocalDateTime#plusWithOverflow}
   */
  private JulianDateTime plusWithOverflow(JulianDate newDate, long hours, long minutes, long seconds, long nanos, int sign) {
    // 9223372036854775808 long, 2147483648 int
    if ((hours | minutes | seconds | nanos) == 0) {
      return with(newDate, time);
    }
    long totDays = nanos / NANOS_PER_DAY +             //   max/24*60*60*1B
      seconds / SECONDS_PER_DAY +                //   max/24*60*60
      minutes / MINUTES_PER_DAY +                //   max/24*60
      hours / HOURS_PER_DAY;                     //   max/24
    totDays *= sign;                                   // total max*0.4237...
    long totNanos = nanos % NANOS_PER_DAY +                    //   max  86400000000000
      (seconds % SECONDS_PER_DAY) * NANOS_PER_SECOND +   //   max  86400000000000
      (minutes % MINUTES_PER_DAY) * NANOS_PER_MINUTE +   //   max  86400000000000
      (hours % HOURS_PER_DAY) * NANOS_PER_HOUR;          //   max  86400000000000
    long curNoD = time.toNanoOfDay();                       //   max  86400000000000
    totNanos = totNanos * sign + curNoD;                    // total 432000000000000
    totDays += Math.floorDiv(totNanos, NANOS_PER_DAY);
    long newNoD = Math.floorMod(totNanos, NANOS_PER_DAY);
    LocalTime newTime = (newNoD == curNoD ? time : LocalTime.ofNanoOfDay(newNoD));
    return with(newDate.plus(totDays , ChronoUnit.DAYS), newTime);
  }


  /**
   * see {@link LocalDateTime#until(Temporal, TemporalUnit)}
   */
  @Override
  public long until(Temporal endExclusive, TemporalUnit unit) {
    JulianDateTime end = JulianDateTime.from(endExclusive);
    if (unit instanceof ChronoUnit) {
      if (unit.isTimeBased()) {
        // long amount = date.daysUntil(end.date);
        long amount = date.until(end.date).get(ChronoUnit.DAYS);
        if (amount == 0) {
          return time.until(end.time, unit);
        }
        long timePart = end.time.toNanoOfDay() - time.toNanoOfDay();
        if (amount > 0) {
          amount--;  // safe
          timePart += NANOS_PER_DAY;  // safe
        }
        else {
          amount++;  // safe
          timePart -= NANOS_PER_DAY;  // safe
        }
        switch ((ChronoUnit) unit) {
          case NANOS:
            amount = Math.multiplyExact(amount, NANOS_PER_DAY);
            break;
          case MICROS:
            amount = Math.multiplyExact(amount, MICROS_PER_DAY);
            timePart = timePart / 1000;
            break;
          case MILLIS:
            amount = Math.multiplyExact(amount, MILLIS_PER_DAY);
            timePart = timePart / 1_000_000;
            break;
          case SECONDS:
            amount = Math.multiplyExact(amount, SECONDS_PER_DAY);
            timePart = timePart / NANOS_PER_SECOND;
            break;
          case MINUTES:
            amount = Math.multiplyExact(amount, MINUTES_PER_DAY);
            timePart = timePart / NANOS_PER_MINUTE;
            break;
          case HOURS:
            amount = Math.multiplyExact(amount, HOURS_PER_DAY);
            timePart = timePart / NANOS_PER_HOUR;
            break;
          case HALF_DAYS:
            amount = Math.multiplyExact(amount, 2);
            timePart = timePart / (NANOS_PER_HOUR * 12);
            break;
        }
        return Math.addExact(amount, timePart);
      }
      JulianDate endDate = end.date;
      if (endDate.isAfter(date) && end.time.isBefore(time)) {
        // endDate = endDate.minusDays(1);
        endDate = endDate.minus(1 , ChronoUnit.DAYS);
      }
      else if (endDate.isBefore(date) && end.time.isAfter(time)) {
        // endDate = endDate.plusDays(1);
        endDate = endDate.plus(1 , ChronoUnit.DAYS);
      }
      return date.until(endDate, unit);
    }
    return unit.between(this, end);
  }

  @Override
  public ZonedJulianDateTime atZone(ZoneId zone) {
    return ZonedJulianDateTime.of(this, zone);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o)
      return true;
    if (!(o instanceof JulianDateTime))
      return false;
    JulianDateTime that = (JulianDateTime) o;
    return Objects.equals(date, that.date) && Objects.equals(time, that.time);
  }

  @Override
  public int hashCode() {
    return Objects.hash(date, time);
  }


  @Override
  public String toString() {
    return "[JulianDateTime " + "date=" + date + ", time=" + time + ']';
  }
}

File: src/main/java/destiny/core/calendar/JulianDateTime.kt
/**
 * Created by smallufo on 2017-01-20.
 */
package destiny.core.calendar

import org.threeten.extra.chrono.JulianDate
import java.io.Serializable
import java.time.*
import java.time.chrono.ChronoLocalDateTime
import java.time.temporal.*
import java.time.temporal.ChronoField.*
import java.util.*

/**
 * reference : [java.time.LocalDateTime]
 */
class JulianDateTime private constructor(private val date: JulianDate, private val time: LocalTime) : Serializable,
  ChronoLocalDateTime<JulianDate> {

  /** 一定為正值  */
  val yearOfEra: Int = date.get(YEAR_OF_ERA)

  /** 與 [LocalDate.year] 一樣，都為 proleptic year . 西元元年=1 , 西元前一年=0  */
  val year: Int = date.get(YEAR)

  val month: Int = date.get(MONTH_OF_YEAR)

  val dayOfMonth: Int = date.get(DAY_OF_MONTH)

  val hour: Int  = time.hour

  val minute: Int = time.minute

  val second: Int = time.second

  val nano: Int = time.nano

  override fun toLocalDate(): JulianDate {
    return date
  }

  override fun toLocalTime(): LocalTime {
    return time
  }

  /**
   * see [java.time.LocalDateTime.isSupported]
   */
  override fun isSupported(field: TemporalField?): Boolean {
    if (field is ChronoField) {
      val f = field as ChronoField?
      return f!!.isDateBased || f.isTimeBased
    }
    return field != null && field.isSupportedBy(this)
  }

  /**
   * see [java.time.LocalDateTime.getLong]
   */
  override fun getLong(field: TemporalField): Long {
    return if (field is ChronoField) {
      if (field.isTimeBased) time.getLong(field) else date.getLong(field)
    } else field.getFrom(this)
  }

  /**
   * see [LocalDateTime.with]
   */
  override fun with(field: TemporalField, newValue: Long): ChronoLocalDateTime<JulianDate> {
    return if (field is ChronoField) {
      if (field.isTimeBased) {
        with(date, time.with(field, newValue))
      } else {
        with(date.with(field, newValue), time)
      }
    } else field.adjustInto(this, newValue)
  }

  /**
   * see [LocalDateTime.with]
   */
  private fun with(newDate: JulianDate, newTime: LocalTime): JulianDateTime {
    return if (date === newDate && time === newTime) {
      this
    } else JulianDateTime(newDate, newTime)
  }

  /**
   * see [LocalDateTime.plus]
   */
  override fun plus(amountToAdd: Long, unit: TemporalUnit): ChronoLocalDateTime<JulianDate> {
    if (unit is ChronoUnit) {
      when (unit) {
        ChronoUnit.NANOS -> return plusNanos(amountToAdd)
        ChronoUnit.MICROS -> return plusDays(amountToAdd / MICROS_PER_DAY).plusNanos(
          amountToAdd % MICROS_PER_DAY * 1000)
        ChronoUnit.MILLIS -> return plusDays(amountToAdd / MILLIS_PER_DAY).plusNanos(
          amountToAdd % MILLIS_PER_DAY * 1000000)
        ChronoUnit.SECONDS -> return plusSeconds(amountToAdd)
        ChronoUnit.MINUTES -> return plusMinutes(amountToAdd)
        ChronoUnit.HOURS -> return plusHours(amountToAdd)
        ChronoUnit.HALF_DAYS -> return plusDays(amountToAdd / 256).plusHours(
          amountToAdd % 256 * 12)  // no overflow (256 is multiple of 2)
      }
      return with(date.plus(amountToAdd, unit), time)
    }
    return unit.addTo(this, amountToAdd)
  }

  override fun minus(amountToSubtract: Long, unit: TemporalUnit): ChronoLocalDateTime<JulianDate> {
    return plus(-amountToSubtract, unit)
  }

  fun plusNanos(nanos: Long): JulianDateTime {
    return plusWithOverflow(date, 0, 0, 0, nanos, 1)
  }

  fun plusSeconds(seconds: Long): JulianDateTime {
    return plusWithOverflow(date, 0, 0, seconds, 0, 1)
  }

  fun plusMinutes(minutes: Long): JulianDateTime {
    return plusWithOverflow(date, 0, minutes, 0, 0, 1)
  }

  fun plusHours(hours: Long): JulianDateTime {
    return plusWithOverflow(date, hours, 0, 0, 0, 1)
  }

  fun plusDays(days: Long): JulianDateTime {
    val newDate = date.plus(days, ChronoUnit.DAYS)
    return with(newDate, time)
  }

  fun plusMonths(months: Long) : JulianDateTime {
    val newDate = date.plus(months, ChronoUnit.MONTHS)
    return with(newDate , time)
  }

  /**
   * see [LocalDateTime.plusWithOverflow]
   */
  private fun plusWithOverflow(newDate: JulianDate,
                               hours: Long,
                               minutes: Long,
                               seconds: Long,
                               nanos: Long,
                               sign: Int): JulianDateTime {
    // 9223372036854775808 long, 2147483648 int
    if (hours or minutes or seconds or nanos == 0L) {
      return with(newDate, time)
    }
    var totDays = nanos / NANOS_PER_DAY +             //   max/24*60*60*1B

      seconds / SECONDS_PER_DAY +                     //   max/24*60*60

      minutes / MINUTES_PER_DAY +                     //   max/24*60

      hours / HOURS_PER_DAY                           //   max/24
    totDays *= sign.toLong()                          // total max*0.4237...
    var totNanos = nanos % NANOS_PER_DAY +            //   max  86400000000000

      seconds % SECONDS_PER_DAY * NANOS_PER_SECOND +  //   max  86400000000000

      minutes % MINUTES_PER_DAY * NANOS_PER_MINUTE +  //   max  86400000000000

      hours % HOURS_PER_DAY * NANOS_PER_HOUR          //   max  86400000000000
    val curNoD = time.toNanoOfDay()                   //   max  86400000000000
    totNanos = totNanos * sign + curNoD               // total 432000000000000
    totDays += Math.floorDiv(totNanos, NANOS_PER_DAY)
    val newNoD = Math.floorMod(totNanos, NANOS_PER_DAY)
    val newTime = if (newNoD == curNoD) time else LocalTime.ofNanoOfDay(newNoD)
    return with(newDate.plus(totDays, ChronoUnit.DAYS), newTime)
  }


  /**
   * see [LocalDateTime.until]
   */
  override fun until(endExclusive: Temporal, unit: TemporalUnit): Long {
    val end = from(endExclusive)
    if (unit is ChronoUnit) {
      if (unit.isTimeBased()) {
        // long amount = date.daysUntil(end.date);
        var amount = date.until(end.date).get(ChronoUnit.DAYS)
        if (amount == 0L) {
          return time.until(end.time, unit)
        }
        var timePart = end.time.toNanoOfDay() - time.toNanoOfDay()
        if (amount > 0) {
          amount--  // safe
          timePart += NANOS_PER_DAY  // safe
        } else {
          amount++  // safe
          timePart -= NANOS_PER_DAY  // safe
        }
        when (unit) {
          ChronoUnit.NANOS -> amount = Math.multiplyExact(amount, NANOS_PER_DAY)
          ChronoUnit.MICROS -> {
            amount = Math.multiplyExact(amount, MICROS_PER_DAY)
            timePart /= 1000
          }
          ChronoUnit.MILLIS -> {
            amount = Math.multiplyExact(amount, MILLIS_PER_DAY)
            timePart /= 1000000
          }
          ChronoUnit.SECONDS -> {
            amount = Math.multiplyExact(amount, SECONDS_PER_DAY.toLong())
            timePart /= NANOS_PER_SECOND
          }
          ChronoUnit.MINUTES -> {
            amount = Math.multiplyExact(amount, MINUTES_PER_DAY.toLong())
            timePart /= NANOS_PER_MINUTE
          }
          ChronoUnit.HOURS -> {
            amount = Math.multiplyExact(amount, HOURS_PER_DAY.toLong())
            timePart /= NANOS_PER_HOUR
          }
          ChronoUnit.HALF_DAYS -> {
            amount = Math.multiplyExact(amount, 2)
            timePart /= (NANOS_PER_HOUR * 12)
          }
        }
        return Math.addExact(amount, timePart)
      }
      var endDate = end.date
      if (endDate.isAfter(date) && end.time.isBefore(time)) {
        // endDate = endDate.minusDays(1);
        endDate = endDate.minus(1, ChronoUnit.DAYS)
      } else if (endDate.isBefore(date) && end.time.isAfter(time)) {
        // endDate = endDate.plusDays(1);
        endDate = endDate.plus(1, ChronoUnit.DAYS)
      }
      return date.until(endDate, unit)
    }
    return unit.between(this, end)
  }

  override fun atZone(zone: ZoneId): ZonedJulianDateTime {
    return ZonedJulianDateTime.of(this, zone)
  }




  override fun toString(): String {
    return "[JulianDateTime date=$date, time=$time]"
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other !is JulianDateTime) return false

    if (date != other.date) return false
    if (time != other.time) return false

    return true
  }

  override fun hashCode(): Int {
    var result = date.hashCode()
    result = 31 * result + time.hashCode()
    return result
  }

  companion object {

    /**
     * Hours per day.
     */
    private const val HOURS_PER_DAY = 24

    /**
     * Minutes per hour.
     */
    private const val MINUTES_PER_HOUR = 60

    /**
     * Minutes per day.
     */
    private const val MINUTES_PER_DAY = MINUTES_PER_HOUR * HOURS_PER_DAY

    /**
     * Seconds per minute.
     */
    private const val SECONDS_PER_MINUTE = 60

    /**
     * Seconds per hour.
     */
    private const val SECONDS_PER_HOUR = SECONDS_PER_MINUTE * MINUTES_PER_HOUR

    /**
     * Seconds per day.
     */
    private const val SECONDS_PER_DAY = SECONDS_PER_HOUR * HOURS_PER_DAY

    /**
     * Milliseconds per day.
     */
    private const val MILLIS_PER_DAY = SECONDS_PER_DAY * 1000L

    /**
     * Microseconds per day.
     */
    private const val MICROS_PER_DAY = SECONDS_PER_DAY * 1000_000L

    /**
     * Nanos per second.
     */
    private const val NANOS_PER_SECOND = 1000_000_000L

    /**
     * Nanos per minute.
     */
    private const val NANOS_PER_MINUTE = NANOS_PER_SECOND * SECONDS_PER_MINUTE

    /**
     * Nanos per hour.
     */
    private const val NANOS_PER_HOUR = NANOS_PER_MINUTE * MINUTES_PER_HOUR

    /**
     * Nanos per day.
     */
    private const val NANOS_PER_DAY = NANOS_PER_HOUR * HOURS_PER_DAY

    fun from(temporal: TemporalAccessor): JulianDateTime {
      if (temporal is JulianDateTime) {
        return temporal
      } else if (temporal is ZonedJulianDateTime) {
        return temporal.toLocalDateTime()
      }
      //    else if (temporal instanceof OffsetDateTime) {
      //      return ((OffsetDateTime) temporal).toLocalDateTime();
      //    }
      try {
        val date = JulianDate.from(temporal)
        val time = LocalTime.from(temporal)
        return JulianDateTime(date, time)
      } catch (ex: DateTimeException) {
        throw DateTimeException(
          "Unable to obtain JulianDateTime from TemporalAccessor: " + temporal + " of type " + temporal.javaClass.name,
          ex)
      }

    }

    fun of(date: JulianDate, time: LocalTime): JulianDateTime {
      Objects.requireNonNull(date, "date")
      Objects.requireNonNull(time, "time")
      return JulianDateTime(date, time)
    }

    /**
     * @param prolepticYear maybe <= 0
     */
    fun of(prolepticYear: Int,
           month: Int,
           dayOfMonth: Int,
           hour: Int,
           minute: Int,
           second: Int,
           nano: Int): JulianDateTime {
      val date = JulianDate.of(prolepticYear, month, dayOfMonth)
      val time = LocalTime.of(hour, minute, second, nano)
      return JulianDateTime(date, time)
    }

    /**
     * @param prolepticYear maybe <= 0
     */
    //@JvmOverloads
    fun of(prolepticYear: Int, month: Int, dayOfMonth: Int, hour: Int, minute: Int, second: Int = 0): JulianDateTime {
      val date = JulianDate.of(prolepticYear, month, dayOfMonth)

      val (first, second1) = TimeTools.splitSecond(second.toDouble())
      val time = LocalTime.of(hour, minute, first, second1)
      return JulianDateTime(date, time)
    }

    /**
     * @param prolepticYear maybe <= 0
     */
    fun of(prolepticYear: Int, month: Month, dayOfMonth: Int, hour: Int, minute: Int): JulianDateTime {
      return of(prolepticYear, month.value, dayOfMonth, hour, minute, 0)
    }

    fun ofEpochSecond(epochSecond: Long, nanoOfSecond: Int, offset: ZoneOffset): JulianDateTime {
      Objects.requireNonNull(offset, "offset")
      NANO_OF_SECOND.checkValidValue(nanoOfSecond.toLong())
      val localSecond = epochSecond + offset.totalSeconds  // overflow caught later
      val localEpochDay = Math.floorDiv(localSecond, SECONDS_PER_DAY.toLong())
      val secsOfDay = Math.floorMod(localSecond, SECONDS_PER_DAY.toLong()).toInt()
      val gDate = LocalDate.ofEpochDay(localEpochDay)
      val date = JulianDate.from(gDate)
      val time = LocalTime.ofNanoOfDay(secsOfDay * NANOS_PER_SECOND + nanoOfSecond)
      return JulianDateTime(date, time)
    }
  }
}
